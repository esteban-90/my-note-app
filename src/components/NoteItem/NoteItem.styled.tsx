import styled from '@emotion/styled'

export const Wrapper = styled.div`
  border-bottom: 1px solid var(--color-border);
  padding: 0.1rem 1rem;
  transition: all 0.2s ease-in-out;
  font-weight: 600;

  @media (hover: hover) {
    &:hover {
      background-color: var(--color-bg);
      cursor: pointer;
    }
  }

  & p {
    font-size: 0.8rem;
    color: var(--color-light);

    & span:first-of-type {
      color: var(--color-gray);
      display: inline-block;
      margin-right: 0.5rem;
    }
  }

  & p,
  h3 {
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
  }

  & h3::first-letter {
    text-transform: uppercase;
  }
`
