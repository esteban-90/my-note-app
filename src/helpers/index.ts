import Filter from 'bad-words-es'

/**
 * Checks if the content of a note has profane vocabulary.
 * @param noteContent The content of the note.
 * @returns true if it does, false otherwise.
 */
export const hasBadWords = (noteContent: string) => {
  const filter = new Filter()
  const isProfane = filter.isProfane(noteContent)
  return isProfane
}

/**
 * Retrieves the title of a note by splitting its content
 * until the first line break, or the first six words.
 * @param noteContent The content of the note.
 * @returns The resulting title.
 */
export const getTitle = (noteContent: string) => {
  let [title] = noteContent.split('\n')
  const splitted = title.split(' ')
  if (splitted.length > 6) title = splitted.slice(0, 6).join(' ')
  return title
}

/**
 * Formats the date of creation of a note according
 * to the current language locale set in the app.
 * @param noteCreation The date of creation, in ISO format.
 * @param locale The language locale.
 * @returns The formatted date.
 */
export const getDate = (noteCreation: string, locale: 'en' | 'es' = 'en') => {
  return new Date(noteCreation).toLocaleString(locale, { dateStyle: 'long', timeStyle: 'medium' })
}

/**
 * Retrieves the body of a note by removing its title and
 * getting the remaining words.
 * @param noteContent The content of the note.
 * @returns The resulting body.
 */
export const getBody = (noteContent: string) => {
  const title = getTitle(noteContent)
  const body = noteContent.replaceAll('\n', ' ').replaceAll(title, '')
  return body
}
