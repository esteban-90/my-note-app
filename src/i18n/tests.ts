import type { Resource } from 'i18next'
import { createInstance } from 'i18next'
import { initReactI18next } from 'react-i18next'
import { ns, supportedLngs } from './app'

const resources: Resource = {}
export const i18nInstance = createInstance()

void (async () => {
  for (const n of ns) {
    for (const lng of supportedLngs) {
      resources[lng] ??= {}
      resources[lng] = {
        ...resources[lng],
        [n]: await import(`../../public/translations/${lng}/${n}.json`),
      }
    }
  }
})()

i18nInstance.use(initReactI18next).init({
  ns,
  supportedLngs,
  fallbackLng: supportedLngs[0],
  lng: supportedLngs[0],
  react: { useSuspense: false },
  resources,
})
