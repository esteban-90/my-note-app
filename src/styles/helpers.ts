import { css } from '@emotion/react'

export const bgTransition = css`
  transition: background-color 0.7s;
`

export const centered = css`
  display: flex;
  align-items: center;
  justify-content: center;
`
