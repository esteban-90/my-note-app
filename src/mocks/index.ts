import type { Note, Language, Theme } from '@/types'
import uniqid from 'uniqid'
import { makeKey } from '@/helpers'

let seconds = 0

export const makeMockNote = (content: string): Note => {
  return {
    id: uniqid('note-'),
    content,
    createdAt: new Date(1995, 11, 17, 3, 24, seconds++).toISOString(),
  }
}

const makeMockNotes = (quantity = 1): Note[] => {
  return new Array(quantity).fill(null).map((_, index) => makeMockNote(`Hello test ${index + 1}\nTesting ${index + 1}`))
}

export const mockNotes = makeMockNotes(5)

export const mockNoteKey = makeKey('notes')
export const mockLanguageKey = makeKey('language')
export const mockThemeKey = makeKey('theme')

export const mockENLanguage: Language = 'en-US'
export const mockESLanguage: Language = 'es-PY'

export const mockDayTheme: Theme = 'day'
export const mockNightTheme: Theme = 'night'

export const formatMockDate = (date: string, locale: Language = 'en-US') => {
  return new Date(date).toLocaleString(locale, { dateStyle: 'long', timeStyle: 'medium' })
}
