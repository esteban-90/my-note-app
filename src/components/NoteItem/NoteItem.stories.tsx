import type { ComponentStory, ComponentMeta } from '@storybook/react'
import { mockNotes } from '@/mocks'
import { NoteItem } from './NoteItem'

export default {
  title: 'Components/Note Item',
  component: NoteItem,
  args: mockNotes[0],
} as ComponentMeta<typeof NoteItem>

const Template: ComponentStory<typeof NoteItem> = (args) => <NoteItem {...args} />

export const NoteItemExample = Template.bind({})
