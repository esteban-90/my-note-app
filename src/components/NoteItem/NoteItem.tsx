import type { FC } from 'react'
import type { Note } from '@/types'
import { Link } from 'react-router-dom'
import { useTranslation } from 'react-i18next'
import { getTitle, getDate, getBody } from '@/helpers'
import { Wrapper } from '@styled/NoteItem'

export const NoteItem: FC<Note> = ({ id, content, createdAt }) => {
  const {
    i18n: { resolvedLanguage },
  } = useTranslation()

  return (
    <Link to={`notes/${id}`} data-testid={id} role='link'>
      <Wrapper>
        <h3>{getTitle(content)}</h3>
        <p>
          <span>{getDate(createdAt, resolvedLanguage)}</span>
          <span>{getBody(content)}</span>
        </p>
      </Wrapper>
    </Link>
  )
}
