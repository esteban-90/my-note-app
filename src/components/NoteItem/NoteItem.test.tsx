import { MemoryRouter } from 'react-router-dom'
import { getTitle, getDate, getBody } from '@/helpers'
import { mockNote } from '@/mocks'
import { screen, customRender } from '@/tests'
import { NoteItem } from './NoteItem'

describe(NoteItem.name, () => {
  let asFragment: () => DocumentFragment

  const UI = (
    <MemoryRouter>
      <NoteItem {...mockNote} />
    </MemoryRouter>
  )

  beforeEach(() => {
    ;({ asFragment } = customRender(UI))
  })

  it('should match snapshot', () => {
    expect(asFragment()).toMatchSnapshot()
  })

  describe(`should render ${NoteItem.name} component with its:`, () => {
    const { content, createdAt } = mockNote

    it('title', () => {
      const noteTitle = getTitle(content)
      const noteTitleElement = screen.getByRole('heading', { name: new RegExp(noteTitle, 'i') })
      expect(noteTitleElement).toBeInTheDocument()
    })

    it('date', () => {
      const noteDate = getDate(createdAt)
      const noteDateElement = screen.getByText(noteDate)
      expect(noteDateElement).toBeInTheDocument()
    })

    it('body', () => {
      const noteBody = getBody(content)
      const noteBodyElement = screen.getByText(noteBody, { trim: false })
      expect(noteBodyElement).toBeInTheDocument()
    })
  })
})
