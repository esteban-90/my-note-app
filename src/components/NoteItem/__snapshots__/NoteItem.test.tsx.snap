// Vitest Snapshot v1

exports[`NOTE ITEM: > should match snapshot 1`] = `
<DocumentFragment>
  <a
    data-testid="note-8d5lli23glblj7ppy"
    href="/notes/note-8d5lli23glblj7ppy"
    role="link"
  >
    <div
      class="css-1ki75lg"
    >
      <h3>
        Testing
      </h3>
      <p>
        <span>
          December 17, 1995 at 3:24:00 AM
        </span>
        <span>
           Note
        </span>
      </p>
    </div>
  </a>
</DocumentFragment>
`;
